### PR TITLE
ADD: new data format for add_data endpoint

### PR DESCRIPTION
**Pull request by madisonmay**
Reviewers: 
----------

- [ ] @mention

What's the purpose of the PR?
-----------------------------

Description of changes
----------------------

Notes to reviewers
------------------

Keep in mind while reviewing code:
----------------------------------
- Is relevant code tested?
- Are added functions/methods documented?
- Separation of concerns (SOC)
- Don't repeat yourself (DRY) 
- Limit the number of positional arguments
- Is function/method length reasonable?
- Can code be broken down into smaller components?
- Where do added functions/methods belong?
- Are variable names descriptive?
- Are errors handled appropriately?
- Can logic be simplified?
- Does the code make sense in context? (expand the diff)

### CODE REVIEW

File: indicoio/custom/custom.py [hunk 1]
```diff
@@ -46,16 +102,19 @@ def add_data(self, data, cloud=None, batch=False, api_key=None, version=None, **
...
-        if batch:
-            data = map(list, data)
-            X, Y = zip(*data)
-            X = image_preprocess(X, batch=batch)
-            # must type cast map obj to list for python3 compatability
-            data = list(map(list, zip(X, Y)))
-        else:
-            data = list(data)
-            data[0] = image_preprocess(data[0], batch=batch)
+        batch = isinstance(data[0], (list, tuple, dict))
```
Comment (sihrc) [inline indicoio/custom/custom.py]: data is not gauranteed to have at least length 1 here is it?

File: indicoio/custom/custom.py [hunk 2]
```diff
@@ -1,11 +1,67 @@
...
+                    "formatted as a list of `[data, target]` pairs."
+                )
+        if isinstance(example, dict):
+            try:
+                xs[idx] = example['data']
+                ys[idx] = example['target']
+                metadata[idx] = example.get('metadata', {})
+            except KeyError:
+                raise IndicoError(
+                    "Invalid input data.  Please ensure input data is "
```
Comment (sihrc) [inline indicoio/custom/custom.py]: ... multiline string?

File: indicoio/custom/custom.py [hunk 3]
```diff
@@ -1,11 +1,67 @@
...
+from itertools import izip
 
 from ..utils.api import api_handler
 from ..utils.decorators import detect_batch
 from ..utils.image import image_preprocess
 from ..utils.errors import IndicoError
 
 
+
+def _unpack_data(data):
```
Comment (sihrc) [inline indicoio/custom/custom.py]: maybe break down into smaller functions?

```
def _unpack_data(data):
    """
    Break Xs, Ys, and metadata out into separate lists for data preprocessing.
    Run basic data validation.
    """
    if not data:
      return [], [], []
    if isinstance(data[0], (list, tuple)):
      return _unpack_list(data)
    elif isinstance(data[0], dict):
      return _unpack_dict(data)
    else:
    raise IndicoError(
      "Invalid data type for add_data. Accepted types are list, tuples, and dicts"
    )

def _unpack_list(data):
    xs = [None] * len(data)
    ys = [None] * len(data)
    for idx, example in enumerate(data):
        try:
            xs[idx] = example[0]
            ys[idx] = example[1]
        except IndexError:
            raise IndicoError(
                "Invalid input data.  Please ensure input data is " + \
                "formatted as a list of `[data, target]` pairs."
            )
    return xs, ys

  def _unpack_dict(data):
     xs = [None] * len(data)
     ys = [None] * len(data)
     metadata = [None] * len(data)
     for idx, example in enumerate(data):
        try:
            xs[idx] = example['data']
            ys[idx] = example['target']
            metadata[idx] = example.get('metadata', {})
        except KeyError:
            raise IndicoError(
                "Invalid input data.  Please ensure input data is " + \
                "formatted as a list of dicts with `data` and `target` keys"
            )
    return xs, ys, metadata
```

Comment (madisonmay) [inline indicoio/custom/custom.py]: Yup, this looks great.

File: indicoio/custom/custom.py (hunk 1 above)
Comment (madisonmay) [inline indicoio/custom/custom.py]: Technically we should do a check to make sure len(data) is not 0.  I'll add that in.

**Pull request closed without merging**